lib/screens/v2/widget_instance/alert.ex

```diff
@@ -2,85 +2,210 @@ defmodule Screens.V2.WidgetInstance.Alert do
+      active_routes: t.active_routes_at_stop,
+      screen: t.screen
     }
+
+    informed_entities = informed_entities(t)
+
+    informed_zones_set =
+      informed_entities
+      |> Enum.flat_map(&informed_entity_to_zone(&1, location_context))
+      |> MapSet.new()
+
+    import MapSet, only: [equal?: 2, new: 1]
+
+    cond do
+      equal?(informed_zones_set, new([:upstream])) -> :upstream
+      equal?(informed_zones_set, new([:downstream])) -> :downstream
+      equal?(informed_zones_set, new([:home_stop])) -> :inside
+      equal?(informed_zones_set, new([:upstream, :home_stop, :downstream])) -> :inside
+      equal?(informed_zones_set, new([:upstream, :home_stop])) -> :boundary_upstream
+      equal?(informed_zones_set, new([:home_stop, :downstream])) -> :boundary_downstream
+      true -> :elsewhere
+    end
   end
 
-  def slot_names(%T{}), do: ~w[medium_left medium_right]a
+  @spec effect(t()) :: Alert.effect()
+  def effect(%__MODULE__{alert: %Alert{effect: effect}}), do: effect
 
-  def widget_type(%T{}), do: :alert
+  @spec informed_entity_to_zone(Alert.informed_entity(), map()) ::
+          list(:upstream | :home_stop | :downstream)
+  defp informed_entity_to_zone(informed_entity, location_context)
 
-  defp active_priority(%T{alert: alert}, alert_active?) do
-    if alert_active?.(alert), do: 0, else: 1
+  # All values nil
+  defp informed_entity_to_zone(%{stop: nil, route: nil, route_type: nil}, _location_context) do
+    []
   end
 
-  defp informed_entity_priority(%T{
-         screen: :ok,
-         alert: %Alert{informed_entities: _informed_entities}
-       }) do
-    0
+  # Only route type is not nil--this is the only time we consider route type,
+  # since it's implied by other values when they are not nil
+  defp informed_entity_to_zone(%{stop: nil, route: nil, route_type: route_type_id}, context) do
+    route_type = RouteType.from_id(route_type_id)
+
+    case {route_type, context.screen.app_id} do
+      {:bus, :bus_shelter_v2} -> [:upstream, :home_stop, :downstream]
+      {:bus, :bus_eink_v2} -> [:upstream, :home_stop, :downstream]
+      {:light_rail, :gl_eink_v2} -> [:upstream, :home_stop, :downstream]
+      _ -> []
+    end
   end
 
-  for {effect, priority} <- @effect_priorities do
-    defp effect_priority(%T{alert: %Alert{effect: unquote(effect)}}) do
-      unquote(priority)
+  # Only stop is not nil (route type ignored)
+  defp informed_entity_to_zone(%{stop: stop, route: nil}, context) do
+    cond do
+      stop == context.home_stop -> [:home_stop]
+      stop in context.upstream_stops -> [:upstream]
+      stop in context.downstream_stops -> [:downstream]
+      true -> []
     end
   end
 
-  defp effect_priority(_), do: 100
+  # Only route is not nil (route type ignored)
+  defp informed_entity_to_zone(%{stop: nil, route: route}, context) do
+    # Should we consider whether the stop is a terminus?
+    # If so, an alert that informs a route through the stop might not always be all 3 zones.
```

Comment:
I don't think this would change priority ranking, but it could potentially matter to `slot_names` on Green Line e-Ink screens, where we check whether the home stop is inside a suspension or shuttle (not on the boundary). In this case, we'd want a whole-route alert to be full-screen -- I think that's consistent with what you have here already.